From adc18d6270955cd21ad4d3697d48872c5881236d Mon Sep 17 00:00:00 2001
From: Veaceslav Falico <vfalico@redhat.com>
Date: Thu, 21 Aug 2014 11:28:28 -0400
Subject: [ethernet] netxen: Print ULA information

Message-id: <1408620515-703-4-git-send-email-vfalico@redhat.com>
Patchwork-id: 88512
O-Subject: [PATCH RHEL7 03/10] netxen_nic: Print ULA information
Bugzilla: 1089356
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

RHBZ#1089356

Testing:
Locally and by QLogic (which supplied the patch)

Description:
Backport of the following commit:

commit 5e7856b31d2848195981880b7f7b8857bd89c967
Author: Shahed Shaikh <shahed.shaikh@qlogic.com>
Date:   Fri Sep 27 01:42:26 2013 -0400

    netxen_nic: Print ULA information

    This patch reads CAMRAM(0x178) where FW writes a key for ULA and non-ULA
    adapter and based on the key, driver logs the message.

    Signed-off-by: Shahed Shaikh <shahed.shaikh@qlogic.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/netxen/netxen_nic_hdr.h b/drivers/net/ethernet/qlogic/netxen/netxen_nic_hdr.h
index 32c7906..0c64c82 100644
--- a/drivers/net/ethernet/qlogic/netxen/netxen_nic_hdr.h
+++ b/drivers/net/ethernet/qlogic/netxen/netxen_nic_hdr.h
@@ -958,6 +958,7 @@ enum {
 #define NETXEN_PEG_HALT_STATUS2  (NETXEN_CAM_RAM(0xac))
 #define NX_CRB_DEV_REF_COUNT  (NETXEN_CAM_RAM(0x138))
 #define NX_CRB_DEV_STATE  (NETXEN_CAM_RAM(0x140))
+#define NETXEN_ULA_KEY   (NETXEN_CAM_RAM(0x178))
 
 /* MiniDIMM related macros */
 #define NETXEN_DIMM_CAPABILITY  (NETXEN_CAM_RAM(0x258))
diff --git a/drivers/net/ethernet/qlogic/netxen/netxen_nic_main.c b/drivers/net/ethernet/qlogic/netxen/netxen_nic_main.c
index 5a119a3..7460d58 100644
--- a/drivers/net/ethernet/qlogic/netxen/netxen_nic_main.c
+++ b/drivers/net/ethernet/qlogic/netxen/netxen_nic_main.c
@@ -1415,6 +1415,32 @@ netxen_setup_netdev(struct netxen_adapter *adapter,
  return 0;
 }
 
+#define NETXEN_ULA_ADAPTER_KEY  (0xdaddad01)
+#define NETXEN_NON_ULA_ADAPTER_KEY (0xdaddad00)
+
+static void netxen_read_ula_info(struct netxen_adapter *adapter)
+{
+ u32 temp;
+
+ /* Print ULA info only once for an adapter */
+ if (adapter->portnum != 0)
+  return;
+
+ temp = NXRD32(adapter, NETXEN_ULA_KEY);
+ switch (temp) {
+ case NETXEN_ULA_ADAPTER_KEY:
+  dev_info(&adapter->pdev->dev, "ULA adapter");
+  break;
+ case NETXEN_NON_ULA_ADAPTER_KEY:
+  dev_info(&adapter->pdev->dev, "non ULA adapter");
+  break;
+ default:
+  break;
+ }
+
+ return;
+}
+
 #ifdef CONFIG_PCIEAER
 static void netxen_mask_aer_correctable(struct netxen_adapter *adapter)
 {
@@ -1561,6 +1587,8 @@ netxen_nic_probe(struct pci_dev *pdev, const struct pci_device_id *ent)
   goto err_out_disable_msi;
  }
 
+ netxen_read_ula_info(adapter);
+
  err = netxen_setup_netdev(adapter, netdev);
  if (err)
   goto err_out_disable_msi;
-- 
1.7.1